docs/proposals/autoscaling-pvcs.md

```diff
@@ -0,0 +1,431 @@
+
+- Monitor volume usage against configurable thresholds.
+- Check if the storage class of the identified PVC supports volume expansion.
+- Automatically increase PVC size when utilization exceeds thresholds.
+- Support for percentage-based and absolute minimum step increases.
+- Cooldown periods to prevent rapid successive scaling operations.
+- Configurable pod restart strategies (manual or automatic).
+
+#### Capacity Constraints
+
+`minCapacity` and `maxCapacity` ensure PVCs stay within reasonable bounds, preventing excessive costs or insufficient storage.
+
+#### Observability and Monitoring
+
+The autoscaler exposes Prometheus metrics for monitoring and alerting:
+- `pvc_autoscaler_resized_total` - Total number of times a PVC has been resized (labeled by namespace, PVC name and owning PVCA name)
+- `pvc_autoscaler_threshold_reached_total` - Total number of times the utilization threshold has been reached (labeled by namespace, PVC name, owning PVCA name and reason: space or inodes)
+- `pvc_autoscaler_max_capacity_reached_total` - Total number of times the max capacity has been reached for a PVC (labeled by namespace, PVC name and owning PVCA name)
+- `pvc_autoscaler_skipped_total` - Total number of times a PVC has been skipped from reconciliation (labeled by namespace, PVC name, owning PVCA name, and reason)
+
+These metrics enable operators to track autoscaling behavior, identify capacity planning needs, and set up alerts for when PVCs reach maximum capacity.
+
+More metrics can be added in the future.
+
+### Scaling Algorithm
+
+The PVC autoscaler employs a **simple threshold-based scaling approach** that makes scaling decisions based on current utilization metrics:
+
+#### Scale-Up Decision Logic
+
+1. Query current volume usage (disk space and inodes) metrics:
+   - `kubelet_volume_stats_available_bytes` - Available bytes on the volume
+   - `kubelet_volume_stats_capacity_bytes` - Total capacity of the volume
+   - `kubelet_volume_stats_inodes_free` - Number of free inodes
+   - `kubelet_volume_stats_inodes` - Total number of inodes
+2. Calculate current utilization percentage for both space and inodes
+3. If storage or inodes usage exceeds `utilizationThresholdPercent` (e.g., 80%), trigger a scale-up operation
+4. Calculate new size: `max(currentSize * (1 + stepPercent/100), currentSize + minStep)`
+6. Respect cooldown periods to prevent thrashing
+7. Cap at `maxCapacity` for scaling up, and at `minCapacity` for downscaling
+
+#### Downscaling (Potential Future Development)
+
+Downscaling PVCs is significantly more complex than upscaling because it requires:
+- Safe data migration to smaller volumes.
+- Coordination with application downtime or maintenance windows.
+- Risk mitigation for potential data loss.
+- Downscaling of some PVCs could result in a size smaller than the one specified in the workload controller that manages the PVC. Depending on the controller, it might try to scale up the PVC back to its original size.
```

Comment:
```suggestion
- Downscaling PVCs could result in a size smaller than the one specified in the workload controller that manages the PVC. Depending on the controller, it might try to scale up the PVC back to its original size.
```